internal/testrunner/runners/system/tester.go

```diff
@@ -1583,6 +1535,118 @@ func (r *tester) createServiceStateDir() error {
 	return nil
 }
 
+func (r *tester) waitForDocs(ctx context.Context, config *testConfig, dataStream string) (*hits, error) {
+	// Use custom timeout if the service can't collect data immediately.
+	waitForDataTimeout := waitForDataDefaultTimeout
+	if config.WaitForDataTimeout > 0 {
+		waitForDataTimeout = config.WaitForDataTimeout
+	}
+
+	if config.Assert.HitCount > elasticsearchQuerySize {
+		return nil, fmt.Errorf("invalid value for assert.hit_count (%d): it must be lower of the maximum query size (%d)", config.Assert.HitCount, elasticsearchQuerySize)
+	}
+
+	if config.Assert.MinCount > elasticsearchQuerySize {
+		return nil, fmt.Errorf("invalid value for assert.min_count (%d): it must be lower of the maximum query size (%d)", config.Assert.MinCount, elasticsearchQuerySize)
+	}
+
+	// (TODO in future) Optionally exercise service to generate load.
+	logger.Debugf("checking for expected data in data stream (%s)...", waitForDataTimeout)
+	var hits *hits
+	oldHits := 0
+	foundFields := map[string]any{}
+	passed, waitErr := wait.UntilTrue(ctx, func(ctx context.Context) (bool, error) {
+		var err error
+		hits, err = r.getDocs(ctx, dataStream)
+		if err != nil {
+			return false, err
+		}
+
+		defer func() {
+			oldHits = hits.size()
+		}()
+
+		if r.checkFailureStore {
+			failureStore, err := r.getFailureStoreDocs(ctx, dataStream)
+			if err != nil {
+				return false, fmt.Errorf("failed to check failure store: %w", err)
+			}
+			if n := len(failureStore); n > 0 {
+				// Interrupt loop earlier if there are failures in the document store.
+				logger.Debugf("Found %d hits in the failure store for %s", len(failureStore), dataStream)
+				return true, nil
+			}
+		}
+
+		assertHitCount := func() bool {
+			if config.Assert.HitCount == 0 {
+				// not enabled
+				return true
+			}
+			if hits.size() < config.Assert.HitCount {
+				return false
+			}
+
+			ret := hits.size() == oldHits
+			if !ret {
+				time.Sleep(4 * time.Second)
+			}
```

Comment:
Should we also add some wait time if this assertion is satisfied but not the others?

Reply:
Not sure I follow this. Why should it be needed to wait when `ret` is true?

In those packages where just this assertion is set, the behavior would be the same now. 
As it is, this assertion required that two consecutive times the hits size is the same once `hits.size() >= HitCount`, waiting 4 seconds when the hits size is not the same as the previous iteration.

In those packages where just this assertion is set, the behavior would be the same now. 

The issue would be when there is assertion is set along with `min_count` or `fields_present`. That time.Sleep would delay the check of the other assertions. But probably this would not cause too much trouble.

Reply:
> Why should it be needed to wait when `ret` is true?

My comment goes more in the line of why do we need to wait four seconds when `ret` is false, and if we do, why we don't need to wait when `ret` is true and some of the other conditions is false.

At least for the min count it looks like we could be interested on the same retry patterns.

Reply:
My assumption is that in the assertion of `hit_count` once the number of hits is equal or great than `config.Assert.HitCount`, that waiting time would be to ensure that there are no more documents ingested. For me, it's like an heuristic that ingestion of documents has finished.

In case of true, it would be considered/assumed that there won't be new messages, and it can be considered true that assertion (so other assertions can be validated).

In case of `min_count`, I think it would not be needed since it is just required to be greater or equal than the given number.

Would that make sense?

I think I'd keep it as it is here by now.

Reply:
Ok :+1: